### PR TITLE
API: Update `lib.index_tricks` namespace

### PR DESCRIPTION
**Pull request by mtsokol**
Relevant issue https://github.com/numpy/numpy/issues/24507

Hi @rgommers @ngoldbaum,

This PR moves `lib.index_tricks` module to a private file and ensures that it's public methods are only available through the main namespace. Looking at overall usage of `c_`, `s_` and `r_` (which is a lot as I checked github search) I decided to keep them in the main namespace, therefore there's no need for local `index_tricks` namespace.

### CODE REVIEW

File: numpy/lib/tests/test_index_tricks.py [hunk 1]
```diff
@@ -405,7 +405,7 @@ def test_repeated_input(self):
 
 
 def test_c_():
-    a = np.c_[np.array([[1, 2, 3]]), 0, 0, np.array([[4, 5, 6]])]
```
Comment (mtsokol) [inline numpy/lib/tests/test_index_tricks.py:408]: For consistency with `r_` and `s_` in this file.

**Comment by ngoldbaum** [comment]
The code changes look good to me and the changes to the `lib` namespace look correct:

```
In [5]: diff = set(oldall) - set(np.lib.__all__)

In [6]: diff.issubset(np.__all__)
Out[6]: True

In [7]: diff
Out[7]: 
{'c_',
 'diag_indices',
 'diag_indices_from',
 'fill_diagonal',
 'index_exp',
 'ix_',
 'mgrid',
 'ndenumerate',
 'ndindex',
 'ogrid',
 'r_',
 'ravel_multi_index',
 's_',
 'unravel_index'}
```

I think is good to merge as soon as the conflict is fixed.

**Comment by mtsokol** [comment]
Rebased and solved conflicts!

**Comment by mtsokol** [comment]
> Looking on github search, it seems there are downstream usages of `np.lib.index_tricks.IndexExpression` and `np.lib.index_tricks.nd_grid`. Both I think need to be available for type annotations and `isinstance` checks. I think that indicates that `index_tricks` needs to remain public with only those two symbols remaining inside it, everything else should be available in the main namespace. @rgommers do you agree?

In my opinion both of them should be private. 
For `IndexExpression` the docstring says:
```
.. note::
    Use one of the two predefined instances `index_exp` or `s_`
    rather than directly using `IndexExpression`.
```
where their definitions are:
```python
index_exp = IndexExpression(maketuple=True)
s_ = IndexExpression(maketuple=False)
```
And this is exhaustive for `IndexExpression` initialization.

And same applies to `nd_grid`, looking at docstring:
```
Two instances of `nd_grid` are made available in the NumPy namespace,
`mgrid` and `ogrid`, approximately defined as::

    mgrid = nd_grid(sparse=False)
    ogrid = nd_grid(sparse=True)

Users should use these pre-defined instances instead of using `nd_grid`
directly.
```

**Comment by mtsokol** [comment]
Ah, you mentioned type annotations - then sure, I can keep them.

**Comment by ngoldbaum** [comment]
Yeah, `isinstance` checks too.

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @mtsokol! Let's give this a go.

**Comment by rgommers** [comment]
> `nd_grid` is the only one for which I get a lot of hits.

For completing the record: I rechecked those hits, and a large majority are outdated forks of the numpy repo or its docstring contents under `See Also`, which had a usage of `np.lib.index_tricks.nd_grid`. There wasn't anything that looked relevant.

**Pull request merged**